crates/ra_assists/src/assists/early_return.rs

```diff
@@ -77,31 +79,67 @@ pub(crate) fn convert_to_guarded_return(ctx: AssistCtx<impl HirDatabase>) -> Opt
-            &mut new_if_and_then_statements,
-        );
+        let new_block = match pat {
+            None => {
+                // If.
+                let early_expression = &(early_expression.to_owned() + ";");
+                let new_if_expr =
+                    if_indent_level.increase_indent(make::if_expression(&expr, early_expression));
+                let then_block_items = IndentLevel::from(1).decrease_indent(then_block.clone());
+                let end_of_then = then_block_items.syntax().last_child_or_token().unwrap();
+                let end_of_then =
+                    if end_of_then.prev_sibling_or_token().map(|n| n.kind()) == Some(WHITESPACE) {
+                        end_of_then.prev_sibling_or_token().unwrap()
+                    } else {
+                        end_of_then
+                    };
+                let mut new_if_and_then_statements =
+                    new_if_expr.syntax().children_with_tokens().chain(
+                        then_block_items
+                            .syntax()
+                            .children_with_tokens()
+                            .skip(1)
+                            .take_while(|i| *i != end_of_then),
+                    );
+                replace_children(
+                    &parent_block.syntax(),
+                    RangeInclusive::new(
+                        if_expr.clone().syntax().clone().into(),
+                        if_expr.syntax().clone().into(),
+                    ),
+                    &mut new_if_and_then_statements,
+                )
+            }
+            _ => {
+                // If-let.
+                let new_match_expr =
+                    if_indent_level.increase_indent(make::let_match_early(expr, early_expression));
+                let then_block_items = IndentLevel::from(1).decrease_indent(then_block.clone());
+                let end_of_then = then_block_items.syntax().last_child_or_token().unwrap();
+                let end_of_then =
+                    if end_of_then.prev_sibling_or_token().map(|n| n.kind()) == Some(WHITESPACE) {
+                        end_of_then.prev_sibling_or_token().unwrap()
+                    } else {
+                        end_of_then
+                    };
+                let mut then_statements = new_match_expr.syntax().children_with_tokens().chain(
+                    then_block_items
+                        .syntax()
+                        .children_with_tokens()
+                        .skip(1)
+                        .take_while(|i| *i != end_of_then),
+                );
+                let new_block = replace_children(
+                    &parent_block.syntax(),
+                    RangeInclusive::new(
+                        if_expr.clone().syntax().clone().into(),
+                        if_expr.syntax().clone().into(),
+                    ),
+                    &mut then_statements,
+                );
+                new_block
+            }
+        };
```

Comment:
It looks like a lot of code between the two cases here is essentially the same

```rust
                let then_block_items = IndentLevel::from(1).decrease_indent(then_block.clone());
                let end_of_then = then_block_items.syntax().last_child_or_token().unwrap();
                let end_of_then =
                    if end_of_then.prev_sibling_or_token().map(|n| n.kind()) == Some(WHITESPACE) {
                        end_of_then.prev_sibling_or_token().unwrap()
                    } else {
                        end_of_then
                    };
```

```rust
                let then_block_items = IndentLevel::from(1).decrease_indent(then_block.clone());
                let end_of_then = then_block_items.syntax().last_child_or_token().unwrap();
                let end_of_then =
                    if end_of_then.prev_sibling_or_token().map(|n| n.kind()) == Some(WHITESPACE) {
                        end_of_then.prev_sibling_or_token().unwrap()
                    } else {
                        end_of_then
                    };
```


It should be possible to reduce code duplication here by pulling common sections outside of the `match`